### PR TITLE
Make indexing and search options independant

### PR DESCRIPTION
**Pull request by julienbourdeau**
This means that if you turn off indexing, for whatever reasons, your can still search them on your frontend.

I think it's better to not delete the indices when the indexing option is turned off because you might need to turn it off temporarily (maintenance, debugging or so) but you still want your frontend to have the Algolia search bar.

**Related:**
* Previous PR #792
* https://discourse.algolia.com/t/using-main-web-search-on-all-store-locations/697

## Notes
 * I added a filter option to the `runTests.sh` script so we can choose which tests to run while developing 🎉 
 * Default configuration is now re-applied before each tets

### CODE REVIEW

File: tests/QueueTest.php [hunk 1]
```diff
@@ -377,10 +369,6 @@ public function testGetJobs()
 
     public function testHugeJob()
     {
-        // Default value - maxBatchSize = 1000
-        setConfig('algoliasearch/queue/number_of_job_to_run', 10);
```
Comment (JanPetr) [inline tests/QueueTest.php]: Why did you remove those?

File: tests/AbstractIndexingTestCase.php [hunk 2]
```diff
@@ -7,30 +7,45 @@ class AbstractIndexingTestCase extends TestCase
...
+    protected $indexPrefix;
+
+    /**
+     * @var array Default value of general store settings (not related to algolia's extention)
+     */
+    protected $defaultConfig = array(
+        'cataloginventory/options/show_out_of_stock' => '0',
+    );
+
+    const DEFAULT_PRODUCT_COUNT = 86;
```
Comment (JanPetr) [inline tests/AbstractIndexingTestCase.php]: Does it make sense to have it just for products?

File: tests/ProductsIndexingTest.php [hunk 3]
```diff
@@ -91,4 +78,24 @@ public function testDefaultIndexableAttributes()
...
+
+        $productIndexer = new Algolia_Algoliasearch_Model_Indexer_Algolia();
+        $this->processTest($productIndexer, 'products', self::DEFAULT_PRODUCT_COUNT);
+    }
+
+    public function testProductAreSearchableIfIndexingIsDisabled()
+    {
+        setConfig('algoliasearch/credentials/enable_backend', '0');
+
+        $this->processQueryOneProduct();
```
Comment (JanPetr) [inline tests/ProductsIndexingTest.php]: I don't get what this test do? It set's config, but don't index anything and just query one product?

Comment (JanPetr) [inline tests/ProductsIndexingTest.php]: Oh, my bad. I just read the name of the test :)

Comment (JanPetr) [inline tests/ProductsIndexingTest.php]: Why did you extract it to a method anyway, when you don't reuse it? Or do you plan to?

File: tests/AbstractIndexingTestCase.php (hunk 2 above)
Comment (julienbourdeau) [inline tests/AbstractIndexingTestCase.php]: We could add one constant for each but the other ones are only used once. They are specific for the test, this one is what you get with the default config.

File: tests/QueueTest.php (hunk 1 above)
Comment (julienbourdeau) [inline tests/QueueTest.php]: Because they settings the default value, which are set in `setUp()` method. I might have missed something though ^^

File: tests/ProductsIndexingTest.php (hunk 3 above)
Comment (julienbourdeau) [inline tests/ProductsIndexingTest.php]: When I wrote it I thought I would reuse it but in the end I only wrote one test. Should I move it back in `ProductsIndexingTest`?

File: tests/AbstractIndexingTestCase.php (hunk 2 above)
Comment (JanPetr) [inline tests/AbstractIndexingTestCase.php]: I'd move the constant at least to ProductIndexingTest.php as it's relevant only to products and not other children of the class.

File: tests/ProductsIndexingTest.php (hunk 3 above)
Comment (JanPetr) [inline tests/ProductsIndexingTest.php]: Yes please :)

File: tests/QueueTest.php [hunk 4]
```diff
@@ -1,8 +1,6 @@
 <?php
 
-use PHPUnit\Framework\TestCase;
-
-class QueueTest extends TestCase
+class QueueTest extends AbstractIndexingTestCase
```
Comment (JanPetr) [inline tests/QueueTest.php]: Why this class inherits AbstractIndexingTestCase when it's not really testing indexing?
If it just need it's dependencies I think the better approach would be to create AbstractTestCase, put all common stuff (like $indexPrefix) there and remove those stuff form AbstractIndexingTestCase.
Then make AbstractIndexingTestCase inherit from AbstractTestCase.
So we avoid use less code inherited everywhere. WDYT?

File: tests/QueueTest.php (hunk 1 above)
Comment (JanPetr) [inline tests/QueueTest.php]: That's true, but I'd keep it here said explicitly, because the is dependent on this value. And in case you change default values in XML file, you might break the tests or make it false positive.

The same applies to testMaxSingleJobSize() method.

File: tests/bootstrap.php [hunk 5]
```diff
@@ -29,10 +27,18 @@ function resetConfigs($configs = array())
...
 
+    $credentialsShortcuts = array(
+        'credentials/application_id', 'credentials/api_key', 'credentials/search_only_api_key', 'credentials/index_prefix'
+    );
+
     foreach ($xml->default->algoliasearch->children() as $section => $subsections) {
         foreach ($subsections as $subsectionName => $subsection) {
             $shortcut = $section.'/'.$subsectionName;
 
+            if (in_array($shortcut, $credentialsShortcuts)) {
```
Comment (JanPetr) [inline tests/bootstrap.php]: 👍

File: tests/QueueTest.php (hunk 4 above)
Comment (julienbourdeau) [inline tests/QueueTest.php]: Definitely! I'll create `AbstractTestCase` class.

**Comment by JanPetr** [comment]
👍  Feel free to merge! 💯  🎉
